include/miopen/miopen.h

```diff
@@ -4667,6 +4667,106 @@ miopenReduceTensor(miopenHandle_t handle,
 /** @} */
 // CLOSEOUT TensorReduce DOXYGEN GROUP
 
+/**
+ * An miopenProblem will describe a problem for different miopen operations.
+ * For now, this is only used for convolution, but could be used for other
+ * operators in the future(such as GEMM, Pooling, etc)
+ */
+MIOPEN_DECLARE_OBJECT(miopenProblem);
+
+typedef enum
+{
+    miopenProblemDirectionForward        = 0,
+    miopenProblemDirectionBackward       = 1,
+    miopenProblemDirectionBackwardWeight = 2,
+} miopenProblemDirection_t;
+
+// Enums with no meaningful default value should probably have invalid default for fool-proofing.
+typedef enum
+{
+    miopenTensorNameInvalid  = 0,
+    miopenTensorConvolutionX = 1,
+    miopenTensorConvolutionW = 2,
+    miopenTensorConvolutionY = 3,
+} miopenTensorName_t;
+
+typedef enum
+{
+    miopenSearchResultsOrderByTime   = 0,
+    miopenSearchResultsOrderByMemory = 1,
+} miopenSearchResultsOrder_t;
+
+miopenStatus_t miopenCreateConvProblem(miopenProblem_t* problem,
+                                       miopenConvolutionDescriptor_t operatorDesc,
+                                       miopenProblemDirection_t direction);
+
+miopenStatus_t miopenDestroyProblem(miopenProblem_t problem);
+
+// Universal problem should have named tensor fields rather than ordered.
+// Previously, different directions had different input orders, which may lead to confusion.
+// Implicitly ordered fields would also lead to a whole lot of magical ids all over the internals.
+miopenStatus_t miopenSetProblemTensorDescriptor(miopenProblem_t problem,
+                                                miopenTensorName_t name,
+                                                const miopenTensorDescriptor_t descriptor);
+
+/**
+ * The miopenSearchOptions will allow the user to configure how search will be
+ * used. This can include things like exhaustiveSearch, and ordering. In the
+ * future, this can support setting different find mode options.
+ */
+MIOPEN_DECLARE_OBJECT(miopenSearchOptions);
+
+miopenStatus_t miopenCreateSearchOptions(miopenSearchOptions_t* options);
+
+miopenStatus_t miopenDestroySearchOptions(miopenSearchOptions_t options);
+
+miopenStatus_t miopenSetSearchOptionTuning(miopenSearchOptions_t options, int value);
+
+miopenStatus_t miopenSetSearchOptionResultsOrder(miopenSearchOptions_t options,
+                                                 miopenSearchResultsOrder_t value);
+
+miopenStatus_t miopenSetSearchOptionWorkspaceLimit(miopenSearchOptions_t options, size_t value);
+
+MIOPEN_DECLARE_OBJECT(miopenSolution);
+
+/**
+ * Finds solutions to a problem by running different applicable solutions. Memory is automatically
+ * allocated.
+ */
+miopenStatus_t miopenFindSolutions(miopenHandle_t handle,
```

Comment:
All the new API calls implicitly assume the operation to be convolution, however, it is not reflected in its name. I would prefer not to have to replicate all these functions for each op and have an operator argument. Even if that is not acceptable, we should at least have `Convolution` in the API name

Reply:
They don't. What would be executed depends on the problem or solution.